### PR TITLE
add new encryption method

### PR DESCRIPTION
**Pull request by tomaszduda23**
Add support for new version of encryption. It was mentioned https://github.com/tomikaa87/gree-remote/issues/52

### CODE REVIEW

**Comment by RobHofmann** [comment]
Hi, i've just tested this. I'm getting a few errors (using the old encryption):

```
2024-05-26 13:26:46.294 ERROR (MainThread) [homeassistant.components.climate] Error while setting up gree platform for climate
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 356, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/gree/climate.py", line 144, in async_setup_platform
    GreeClimate(hass, name, ip_addr, port, mac_addr, timeout, target_temp_step, temp_sensor_entity_id, lights_entity_id, xfan_entity_id, health_entity_id, powersave_entity_id, sleep_entity_id, eightdegheat_entity_id, air_entity_id, target_temp_entity_id, hvac_modes, fan_modes, swing_modes, preset_modes, auto_xfan, auto_light, encryption_version, encryption_key, uid)
  File "/config/custom_components/gree/climate.py", line 199, in __init__
    self._encryption_key = self.GetDeviceKey().encode("utf8")
                           ^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 302, in GetDeviceKey
    return self.FetchResult(cipher, self._ip_addr, self._port, self._timeout, jsonPayloadToSend)['key']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 286, in FetchResult
    tag = receivedJson['tag']
          ~~~~~~~~~~~~^^^^^^^
KeyError: 'tag'
2024-05-26 13:26:46.379 ERROR (MainThread) [homeassistant.components.climate] Error while setting up gree platform for climate
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 356, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/gree/climate.py", line 144, in async_setup_platform
    GreeClimate(hass, name, ip_addr, port, mac_addr, timeout, target_temp_step, temp_sensor_entity_id, lights_entity_id, xfan_entity_id, health_entity_id, powersave_entity_id, sleep_entity_id, eightdegheat_entity_id, air_entity_id, target_temp_entity_id, hvac_modes, fan_modes, swing_modes, preset_modes, auto_xfan, auto_light, encryption_version, encryption_key, uid)
  File "/config/custom_components/gree/climate.py", line 199, in __init__
    self._encryption_key = self.GetDeviceKey().encode("utf8")
                           ^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 302, in GetDeviceKey
    return self.FetchResult(cipher, self._ip_addr, self._port, self._timeout, jsonPayloadToSend)['key']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 286, in FetchResult
    tag = receivedJson['tag']
          ~~~~~~~~~~~~^^^^^^^
KeyError: 'tag'
2024-05-26 13:26:46.451 ERROR (MainThread) [homeassistant.components.climate] Error while setting up gree platform for climate
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 356, in _async_setup_platform
    await asyncio.shield(awaitable)
  File "/config/custom_components/gree/climate.py", line 144, in async_setup_platform
    GreeClimate(hass, name, ip_addr, port, mac_addr, timeout, target_temp_step, temp_sensor_entity_id, lights_entity_id, xfan_entity_id, health_entity_id, powersave_entity_id, sleep_entity_id, eightdegheat_entity_id, air_entity_id, target_temp_entity_id, hvac_modes, fan_modes, swing_modes, preset_modes, auto_xfan, auto_light, encryption_version, encryption_key, uid)
  File "/config/custom_components/gree/climate.py", line 199, in __init__
    self._encryption_key = self.GetDeviceKey().encode("utf8")
                           ^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 302, in GetDeviceKey
    return self.FetchResult(cipher, self._ip_addr, self._port, self._timeout, jsonPayloadToSend)['key']
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/gree/climate.py", line 286, in FetchResult
    tag = receivedJson['tag']
          ~~~~~~~~~~~~^^^^^^^
KeyError: 'tag'
```

**Comment by tomaszduda23** [comment]
Could you please check again? I have no version 1 unit to tests.

**Comment by RobHofmann** [comment]
> Could you please check again? I have no version 1 unit to tests.

Seems to work now. Few code questions. Let me know :)

**Comment by tomaszduda23** [comment]
Hmmm. I cannot see any questions.

**Comment by RobHofmann** [comment]
image
These 3 comments.

**Comment by tomaszduda23** [comment]
You have to publish it to let others see.

**Comment by RobHofmann** [comment]
i think this should do the trick. can you now see it?

File: custom_components/gree/climate.py [hunk 1]
```diff
@@ -282,9 +301,36 @@ def GetDeviceKey(self):
         jsonPayloadToSend = '{"cid": "app","i": 1,"pack": "' + pack + '","t":"pack","tcid":"' + str(self._mac_addr) + '","uid": 0}'
         return self.FetchResult(cipher, self._ip_addr, self._port, self._timeout, jsonPayloadToSend)['key']
 
+    def GetDeviceKeyGCM(self):
+        _LOGGER.info('Retrieving HVAC encryption key')
+        GENERIC_GREE_DEVICE_KEY = b'{yxAHAY_Lm6pbC/<'
```
[the remark this answers is not in the document]
Comment (tomaszduda23) [reply custom_components/gree/climate.py:306]: It seems like. It is basically the same pattern as before.

[the remark this answers is not in the document]
Comment (RobHofmann) [reply custom_components/gree/climate.py:306]: Interesting, but clear :).

**Comment by tomaszduda23** [comment]
I removed duplicated code. It should look better now.

**Comment by JensVanhooydonck** [comment]
This also works for v1.08 (Vaillant Firmware) 👍

**Comment by RobHofmann** [comment]
Works like a charm! Thanks :)

**Review by RobHofmann — APPROVED** [review]

**Comment by jodewee** [comment]
> This also works for v1.08 (Vaillant Firmware) 👍

Hi @JensVanhooydonck , i have also the same brand, but can't get it to work even with this new release.
Did you also connect the ac to the app first? (i have mine in the gree & ewpe app)
  - platform: gree
    name: GREE_AC_Slaapkamer
    host: 192.168.2.108
    port: 7000
    target_temp_step: 1
    mac: '50:2c:c6:6c:d3:63 '
    encryption_version: 2

**Comment by JensVanhooydonck** [comment]
> > This also works for v1.08 (Vaillant Firmware) 👍
> 
> Hi @JensVanhooydonck , i have also the same brand, but can't get it to work even with this new release. Did you also connect the ac to the app first? (i have mine in the gree & ewpe app)
> 
> * platform: gree
>   name: GREE_AC_Slaapkamer
>   host: 192.168.2.108
>   port: 7000
>   target_temp_step: 1
>   mac: '50:2c:c6:6c:d3:63 '
>   encryption_version: 2

I've added them to the app, and extracted the encryption keys from the database (i'm on a macbook m1 => Opened the Sql-lite databae to get the keys)

```- platform: gree
  name: Airco Office
  host: 192.168.0.147
  port: 7000
  mac: "502cc66cac26"
  target_temp_step: 1
  encryption_key: jRyGnTBiQxF3u8j3
  encryption_version: 2
  uid: 0
```

**Comment by jodewee** [comment]
Hi @JensVanhooydonck 
many, many thanks man, after 2years of struggeling:
it also works without providing the encryption key.
  - platform: gree
    name: GREE_AC_Slaapkamer
    host: 192.168.2.108
    port: 7000
    target_temp_step: 1
    mac: "502cc66cd363"
    encryption_version: 2
    uid: 0

image

**Comment by toughvj** [comment]
@tomaszduda23 powieliliście ten sam błąd, który był w oryginalnym kodzie.

**Comment by jodewee** [comment]
I can confirm it is working flawlessly with:
Bulex vivair multisplit units:

- 20-080MC4NO outdoor unit
- 20-025 NWI (2,5 kW) indoor unit
- 20-035 NWI (3,5 kW) indoor unit

maybe it is possible to add this to the confirmed working units? (they are btw exact the same units as the saunier duval & vaillaint vivair series)

Configured it via the ewpe smart app. (also working in the gree app)

Only configuration needed in ha:

  - platform: gree
    name: ACNAMe
    host: xxx.xxx.xxx.xxx
    port: 7000
    target_temp_step: 1
    mac: "xxxxxxxxxx" 
    encryption_version: 2
    uid: 0

all basic functions are tested & working correctly
image

**Pull request merged**